src/matrix_graph.rs

```diff
@@ -0,0 +1,1147 @@
+use std::marker::PhantomData;
+use std::ops::{Index, IndexMut};
+
+use std::cmp;
+use std::mem;
+
+use ordermap::OrderMap;
+
+use fixedbitset::FixedBitSet;
+
+use {
+    Directed,
+    EdgeType,
+    Outgoing,
+    Undirected,
+    Direction,
+    IntoWeightedEdge,
+};
+
+use graph::NodeIndex as GraphNodeIndex;
+
+use visit::{
+    Data,
+    GetAdjacencyMatrix,
+    GraphBase,
+    GraphProp,
+    IntoEdgeReferences,
+    IntoEdges,
+    IntoNeighbors,
+    IntoNeighborsDirected,
+    IntoNodeIdentifiers,
+    IntoNodeReferences,
+    NodeCount,
+    NodeIndexable,
+    NodeCompactIndexable,
+    Visitable,
+};
+
+pub use graph::IndexType;
+
+// The following types are used to control the max size of the adjacency matrix. Since the maximum
+// size of the matrix vector's is the square of the maximum number of nodes, the number of nodes
+// should be reasonably picked.
+type Ix = u16;
+
+pub type NodeIndex = GraphNodeIndex<Ix>;
+
+#[inline]
+pub fn node_index(ax: usize) -> NodeIndex {
+    debug_assert!(ax.index() <= Ix::max_value() as usize);
+    NodeIndex::new(ax)
+}
+
+#[derive(Clone)]
+pub struct MatrixGraph<N, E, Ty = Directed> {
+    node_adjacencies: Vec<Option<E>>,
+    node_capacity: usize,
+    nodes: IdStorage<N>,
+    nb_edges: usize,
+    ty: PhantomData<Ty>,
+}
+
+pub type DiMatrix<N, E> = MatrixGraph<N, E, Directed>;
+pub type UnMatrix<N, E> = MatrixGraph<N, E, Undirected>;
+
+impl<N, E, Ty: EdgeType> MatrixGraph<N, E, Ty> {
+    /// Create a new `MatrixGraph` with estimated capacity for nodes.
+    pub fn with_nodes(nodes: usize) -> Self {
+        let mut m = MatrixGraph {
+            node_adjacencies: vec![],
+            node_capacity: 0,
+            nodes: IdStorage::with_capacity(nodes),
+            nb_edges: 0,
+            ty: PhantomData,
+        };
+
+        m.extend_capacity_to(nodes);
+
+        m
+    }
+
+    #[inline]
+    fn extend_capacity_to(&mut self, min_node_index: usize) {
+        debug_assert!(min_node_index <= Ix::max_value() as usize);
+        self.node_capacity = extend_linearized_matrix::<Ty, _>(&mut self.node_adjacencies, self.node_capacity, min_node_index);
+    }
+
+    #[inline]
+    fn to_edge_position(&self, a: NodeIndex, b: NodeIndex) -> usize {
+        to_linearized_matrix_position::<Ty>(a.index(), b.index(), self.node_capacity)
+    }
+
+    /// Remove all nodes and edges
+    #[inline]
+    pub fn clear(&mut self) {
+        *self = MatrixGraph::default();
```

Comment:
In a typical datastructure, clear should remove all contents but not drop the capacity. That said, we can wait with fixing this until more whole picture issues are resolved.

Reply:
I've just pushed a `clear()` which should keep the capacity as is. Note that it's `O(V^2)` because the entire adjacency matrix needs to have each of its entries set to `None`, although it does allow much faster node / edge addition afterwards.

Reply:
Right, that's certainly a drawback, but it's expected.